docs/rfcs/099-safekeeper-s3-integration.md

```diff
@@ -0,0 +1,86 @@
+# What
+
+Safe keeper needs to backup WAL to S3
+
+S3 is used across the Neon for a variety of components and sharing code will reduce code maintanance cost 
+
+# Why
+
+In situations when other sources of data are compromised transaction log (WAL) produced by compute becomes the ultimate recovery solution 
+
+see https://github.com/zenithdb/zenith/issues/1169 
+
+
+# Proposal
+
+
+## A common S3 integration
+
+Given that S3 integration is common between both Safekeepers and Page Servers a common library can be built
+The common requirement is an ability for asynchrounous upload of objects to a Remote Storage.
+
+Objects can be represented by a byte stream or by a file, optional compression can be applied on either
+Remote storage shuold have production implementation and test implementations. Production impementation is built to work with S3 API compatible blob storage while test implementation is based on a filesystem
+
+ Commmon operations to be supported:
```

Comment:
As discussed today, I'd like us to be very careful with any garbage collection in last resort storage.
It seems that a GC task would remove (or move to Glacier) objects from S3 (it's the last copy of the data in such case), but I'd rather not make this part of either SK or PS responsibility (one bad move and we can't restore data)

Reply:
Out of curiosity, is S3 blob metadata useful enough to keep in those common operations?

For instance now, on every checkpoint, we have a new Lsm layer file and the metadata file update. When that new file gets uploaded to S3, it could have its updated metadata file stored in the blob metadata.